tariff/ngeso/api.go

```diff
@@ -0,0 +1,139 @@
+
+// ConstructNationalForecastRequest returns a request object to be used when calling the national API.
+func ConstructNationalForecastRequest() *CarbonForecastNationalRequest {
+	return &CarbonForecastNationalRequest{}
+}
+
+// ConstructRegionalForecastByIDRequest returns a validly formatted, fully qualified URI to the forecast valid for the given region.
+func ConstructRegionalForecastByIDRequest(r string) *CarbonForecastRegionalRequest {
+	return &CarbonForecastRegionalRequest{regionid: r}
+}
+
+// ConstructRegionalForecastByPostcodeRequest returns a validly formatted, fully qualified URI to the forecast valid for the given postcode.
+func ConstructRegionalForecastByPostcodeRequest(p string) *CarbonForecastRegionalRequest {
+	return &CarbonForecastRegionalRequest{postcode: p}
+}
+
+type CarbonForecastRequest interface {
+	URI() (string, error)
+	DoRequest(helper *request.Helper) (CarbonForecastResponse, error)
+}
+
+type CarbonForecastNationalRequest struct{}
+
+func (r *CarbonForecastNationalRequest) URI() (string, error) {
+	currentTs := time.Now().UTC()
+	t := currentTs.Format(time.RFC3339)
+	return fmt.Sprintf(ForecastNationalURI, t), nil
+}
+
+func (r *CarbonForecastNationalRequest) DoRequest(client *request.Helper) (res CarbonForecastResponse, err error) {
+	uri, err := r.URI()
+	if err != nil {
+		return
+	}
+	res = &NationalIntensityResult{}
+	err = client.GetJSON(uri, &res)
+	return
+}
+
+type CarbonForecastRegionalRequest struct {
+	regionid string
+	postcode string
+}
+
+func (r *CarbonForecastRegionalRequest) URI() (string, error) {
+	currentTs := time.Now().UTC()
+	t := currentTs.Format(time.RFC3339)
+	if r.regionid != "" {
+		return fmt.Sprintf(ForecastRegionalByIdURI, t, r.regionid), nil
+	} else if r.postcode != "" {
+		return fmt.Sprintf(ForecastRegionalByPostcodeURI, t, r.postcode), nil
+	}
+
+	return "", ErrRegionalRequestInvalidFormat
+}
+
+func (r *CarbonForecastRegionalRequest) DoRequest(client *request.Helper) (res CarbonForecastResponse, err error) {
+	uri, err := r.URI()
+	if err != nil {
+		return
+	}
+	res = &RegionalIntensityResult{}
+	err = client.GetJSON(uri, &res)
+	return
+}
+
+type CarbonForecastResponse interface {
+	Results() []CarbonIntensityForecastEntry
+}
+
+// RegionalIntensityResult is returned by Regional requests. It wraps all data inside a data element.
+// Because that makes sense, and makes all of this SO much easier. /s
+type RegionalIntensityResult struct {
+	Data RegionalIntensityResultData `json:"data"`
+}
+
+func (r RegionalIntensityResult) Results() []CarbonIntensityForecastEntry {
+	return r.Data.Rates
+}
+
+// RegionalIntensityResultData is returned by Regional requests. It includes a bit of extra data.
+type RegionalIntensityResultData struct {
+	RegionId  int                            `json:"regionid"`
+	DNORegion string                         `json:"dnoregion"`
+	ShortName string                         `json:"shortname"`
+	Rates     []CarbonIntensityForecastEntry `json:"data"`
```

Comment:
I usually tr to stick to the json names, but thats personal style.

Reply:
I opted to switch the name as "data" gets nested so many times in this API that it can make it quite difficult to ascertain what's actually at each variable